### PR TITLE
[TE] metric correlation ranking pipeline

### PR DESCRIPTION
**Pull request by apucher**
* algorithm for online correlation of metrics

* support multiple targets and candidates efficiently

* tolerate missing data and partial failure

* RCA pipeline wrapper

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricCorrelationRankingPipeline.java [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+      for(MetricEntity candidate : candidateMetrics) {
+        if(!pctChanges.containsKey(target)) {
+          LOG.warn("No diff data for candidate metric '{}'. Skipping.", candidate.getUrn());
+          continue;
+        }
+
+        DoubleSeries changesCandidate = pctChanges.get(candidate);
+
+        try {
+          double score = Math.abs(changesTarget.corr(changesCandidate)) * target.getScore();
```
Comment (tabaaway) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricCorrelationRankingPipeline.java]: What is target.getScore() here? When will you have multiple target metrics?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricCorrelationRankingPipeline.java [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class MetricCorrelationRankingPipeline extends Pipeline {
```
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricCorrelationRankingPipeline.java]: Could you add some javadoc explaining what this pipeline will do

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricCorrelationRankingPipeline.java [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+  private static final Logger LOG = LoggerFactory.getLogger(MetricCorrelationRankingPipeline.class);
+
+  private static final String PROP_TARGET_INPUT = "targetInput";
```
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricCorrelationRankingPipeline.java]: what does this property mean?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricCorrelationRankingPipeline.java [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+
+  @Override
+  public PipelineResult run(PipelineContext context) {
+    Set<MetricEntity> candidateMetrics = filterNonBaselineContext(context);
+
+    TimeRangeEntity currentRange = TimeRangeEntity.getContextCurrent(context);
+    TimeRangeEntity baselineRange = TimeRangeEntity.getContextBaseline(context);
+
+    Set<MetricEntity> targetMetrics = new HashSet<>();
+    for(Entity e : context.getInputs().get(this.targetInput)) {
```
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricCorrelationRankingPipeline.java]: e-->entity

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricCorrelationRankingPipeline.java [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+    List<RequestContainer> requestList = new ArrayList<>();
+    requestList.addAll(makeRequests(targetMetrics, currentRange, PRE_CURRENT));
+    requestList.addAll(makeRequests(candidateMetrics, currentRange, PRE_CURRENT));
+    requestList.addAll(makeRequests(targetMetrics, baselineRange, PRE_BASELINE));
+    requestList.addAll(makeRequests(candidateMetrics, baselineRange, PRE_BASELINE));
+
+    LOG.info("Requesting {} time series", requestList.size());
+
+    Map<String, RequestContainer> requests = new HashMap<>();
+    for(RequestContainer c : requestList)
```
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricCorrelationRankingPipeline.java]: c--> something verbose

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricCorrelationRankingPipeline.java [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+    List<RequestContainer> requestList = new ArrayList<>();
+    requestList.addAll(makeRequests(targetMetrics, currentRange, PRE_CURRENT));
+    requestList.addAll(makeRequests(candidateMetrics, currentRange, PRE_CURRENT));
+    requestList.addAll(makeRequests(targetMetrics, baselineRange, PRE_BASELINE));
+    requestList.addAll(makeRequests(candidateMetrics, baselineRange, PRE_BASELINE));
+
+    LOG.info("Requesting {} time series", requestList.size());
+
+    Map<String, RequestContainer> requests = new HashMap<>();
+    for(RequestContainer somethingVerbose : requestList) {
```
Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricCorrelationRankingPipeline.java]: 😆

Comment (npawar) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricCorrelationRankingPipeline.java]: No but seriously, it's difficult to follow code with e,t,f,c,df etc.. Please change those

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricCorrelationRankingPipeline.java (hunk 1 above)
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/rootcause/impl/MetricCorrelationRankingPipeline.java]: the current weighting/importance of the target metric. when the user's search context contains multiple metrics (with potentially different weightings)

**Pull request merged**